src/main/java/ch/cern/eam/wshub/core/services/administration/impl/UserGroupMenuServiceImpl.java

```diff
@@ -0,0 +1,321 @@
+            if (node.menuCode.isEmpty()) { // But if no function, then we remove only one path item at the end
+                recursiveMenuSpecification = new UserGroupMenuService.MenuSpecification(String.join("/", Arrays.copyOf(words, words.length - 1)), node.menuCode, node.userGroup);
+            }
+            this.addToMenuHierarchy(context, recursiveMenuSpecification);
+            menuEntries = this.getExtMenuHierarchyAsList(context, recursiveMenuSpecification); //TODO Calling a get two times is not good; fixed by using a HashMap or a tree in a future iteration
+        }
+
+        // Now add leaf item
+        // Decide menu type: if function type is not set, then assume it's a menu and not a function
+        String menuType = this.decideMenuType(node.menuCode, words);
+
+        // Find parent id of new entry/item from previous list
+        GenericMenuEntry parent;
+        if (menuType.equals("M") || menuType.equals("F")) { //TODO better with an enum, even if it's not our design (but might be too cluttered)
+            parent = this.getEntryByPathFromList(menuEntries, Arrays.copyOf(words, words.length - 1));
+        } else { // If to be added is a function, get end of path
+            parent = this.getEntryByPathFromList(menuEntries, words);
+        }
+
+        // With the previous ID found and the menu type determined, fill the request object for both menu item or function item
+        String id = parent != null ? parent.getId() : ""; // Which would be the extMenuCode of the parent..
+        MP6043_AddExtMenus_001 addExtMenus = new MP6043_AddExtMenus_001();
+        ExtMenus extMenus = new ExtMenus();
+        addExtMenus.setExtMenus(extMenus);
+        extMenus.setUSERGROUPID(new USERGROUPID_Type());
+        extMenus.getUSERGROUPID().setUSERGROUPCODE(node.userGroup);
+        extMenus.setFUNCTIONID(new FUNCTIONID_Type());
+        String menuCodeToSend = node.menuCode;
+        if (menuType.equals("M") || menuType.equals("F")) { //TODO better with an enum, even if it's not our design (but might be too cluttered)
+            menuCodeToSend = "BSFOLD"; // And set internal BSFOLD code for menu item
+            extMenus.getFUNCTIONID().setFUNCTIONDESCRIPTION(words[words.length - 1]); // The last item on the path provided (the name of the menu item to add)
+        }
+        extMenus.getFUNCTIONID().setFUNCTIONCODE(menuCodeToSend); // Menu code is function code..
+        extMenus.setEXTMENUPARENT(id);
+        extMenus.setEXTMENUTYPE(menuType);
+        extMenus.setSEQUENCENUMBER(100); //TODO set to a different number depending on position
+        extMenus.setMOBILE("false");
+
+        // With the request object created, perform the add operation
+       tools.performInforOperation(context, inforws::addExtMenusOp, addExtMenus);
+
+       return "OK";
+    }
+
+    private void validateInputNode(MenuSpecification node) {
+        if (node.path == null || node.menuCode == null || node.userGroup == null) {
```

Comment:
As per discussed in person with Goran, this first iteration should be a little bit more safe than powerful. I could quickly change to not be so restricted; could you give an argument why should we be more open here? I would suggest as a solution, instead of modifying this endpoint, creating a new one that accepts a list, which is more programatic-friendly; the idea of the string path is for it to be used more 'manually' by our support team, maybe just exporting a CSV file of paths to be added. However, I believe this point is discussable.

Reply:
As we discussed before, we shouldn't confuse what core functionality is with what user interface should be, nor where (projects) they shall exist